lib/ex_unit/test/ex_unit/pattern_format_test.exs

```diff
@@ -0,0 +1,581 @@
+Code.require_file("../test_helper.exs", __DIR__)
+
+defmodule ExUnit.PatternFormatTest do
+  use ExUnit.Case
+
+  alias ExUnit.{Pattern, PatternFormat}
+
+  test "Two primatives" do
+    # send(self(), 2)
+    # assert_receive 1
+
+    left =
+      quote do
+        1
+      end
+
+    pattern = Pattern.new(left, [], %{})
+    actual = PatternFormat.script(pattern, 2)
+
+    expected = [
+      diff_delete: "2"
+    ]
+
+    assert actual == expected
+  end
+
+  test "message received, map" do
+    # send(self(), %{a: 1, b: 1, d: 2, c: 3})
+    # assert_receive %{a: 2}
+
+    left =
+      quote do
+        %{a: 2}
+      end
+
+    pattern = Pattern.new(left, [], %{})
+    actual = PatternFormat.script(pattern, %{a: 1, b: 1, d: 2, c: 3})
+
+    expected = [
+      "%{",
+      "a: ",
+      {:diff_delete, "1"},
+      ", ",
+      "b: 1",
+      ", ",
+      "c: 3",
+      ", ",
+      "d: 2",
+      "}"
+    ]
+
+    assert actual == expected
+  end
+
+  test "message received, map with mulitble keys" do
+    # send(self(), %{a: 1, b: 1, d: 2, c: 3})
+    # assert_receive %{a: 1, b: 2}
+
+    left = quote do: %{a: 1, b: 2}
+
+    pattern = Pattern.new(left, [], %{})
+    actual = PatternFormat.script(pattern, %{a: 1, b: 1, d: 2, c: 3})
+
+    expected = [
+      "%{",
+      "a: ",
+      "1",
+      ", ",
+      "b: ",
+      {:diff_delete, "1"},
+      ", ",
+      "c: 3",
+      ", ",
+      "d: 2",
+      "}"
+    ]
+
+    assert actual == expected
+  end
+
+  test "message received, map with multible keys don't match" do
+    # send(self(), %{a: 2, b: 1})
+    # assert_receive %{a: 1, b: 2}
+
+    left = quote do: %{a: 1, b: 2}
+
+    pattern = Pattern.new(left, [], %{})
+    actual = PatternFormat.script(pattern, %{a: 2, b: 1})
+
+    expected = ["%{", "a: ", {:diff_delete, "2"}, ", ", "b: ", {:diff_delete, "1"}, "}"]
+    assert actual == expected
+  end
+
+  test "message received, list" do
+    # send(self(), [1, 2, 3])
+    # send(self(), [1, 2])
+    # send(self(), [1, 2, 3, 4, 5])
+    # assert_receive [1, 2, 3, 4]
+
+    left = quote do: [1, 2, 3, 4]
+    pattern = Pattern.new(left, [], %{})
+
+    actual = PatternFormat.script(pattern, [1, 2, 3])
+    expected = ["[", "1", ", ", "2", ", ", "3", ", ", {:diff_insert, "4"}, "]"]
```

Comment:
I just want to note that we were used to this when we first added diffing and it was very confusing and everyone was puzzled. Basically, it is hard for us to see something like `[1, 2, 3, 4]` where `, 4` is green and understand it as "the list is actually only `[1, 2, 3]` and `, 4` is what is missing". That's when we decided for the original diff implementation that we should never change what is printed, we can only annotate it with colors. This means we can show only reds or greens.